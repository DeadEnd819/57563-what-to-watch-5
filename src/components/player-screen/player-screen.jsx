import React, {useState, useEffect, useCallback, useRef} from "react";
import PropTypes from "prop-types";
import {FilmCardType} from "../../prop-types/prop-types";
import {Link} from "react-router-dom";
import {connect} from "react-redux";
import PlayerButton from "../player-button/player-button";
import FullScreenButton from "../full-screen-button/full-screen-button";
import {fetchCurrentFilm} from "../../store/api-actions";
import {getCurrentFilm} from "../../store/selectors";
import {getVideoDuration, getToggleProgress, extend} from "../../utils";
import {AppRoute} from "../../const";

const {FILMS} = AppRoute;

const PlayerScreen = ({id, film, loadDataFilm}) => {
  const {name, backgroundImage, videoLink} = film;
  const videoRef = useRef();

  const [playerState, setPlayerState] = useState({
    isPlaying: true,
    duration: 0,
    progress: 0,
  });

  const {isPlaying, duration, progress} = playerState;
  const timeLeft = getVideoDuration(duration - progress);
  const toggleProgress = getToggleProgress(progress, duration);

  useEffect(() => {
    loadDataFilm(id);
  }, [id]);

  useEffect(() => {
    const video = videoRef.current;

    if (video) {
      video.oncanplay = () => {
        setPlayerState((prevState) => (
          extend(prevState, {duration: Math.floor(video.duration)})
        ));
      };

      video.ontimeupdate = () => {
        setPlayerState((prevState) => (
          extend(prevState, {progress: Math.floor(video.currentTime)})
        ));
      };

      if (isPlaying) {
        video.play();
      } else {
        video.pause();
      }
    }

    return () => {
      if (video) {
        video.oncanplay = null;
        video.ontimeupdate = null;
      }
    };
  }, [isPlaying]);

  const handleTogglePlayButtonClick = useCallback(
      () => {
        setPlayerState((prevState) => (
          extend(prevState, {isPlaying: !isPlaying})
        ));
      }, [isPlaying]
  );

  const handleFullScreenButtonClick = useCallback(
      () => {
        const video = videoRef.current;
        video.requestFullscreen();
      }, [videoRef]
  );

  return (
    <div className="player">
      <video ref={videoRef} src={videoLink} className="player__video" poster={backgroundImage} />

      <Link to={`${FILMS}/${id}`} type="button" className="player__exit">Exit</Link>

      <div className="player__controls">
        <div className="player__controls-row">
          <div className="player__time">
            <progress className="player__progress" value={progress} max={duration}>
            </progress>
            <div className="player__toggler" style={{left: `${toggleProgress}%`}}>Toggler</div>
          </div>
          <div className="player__time-value">{timeLeft}</div>
        </div>

        <div className="player__controls-row">
          <PlayerButton isPlaying={isPlaying} onTogglePlayButtonClick={handleTogglePlayButtonClick}/>
          <div className="player__name">{name}</div>

          <FullScreenButton onFullScreenButtonClick={handleFullScreenButtonClick} />
        </div>
      </div>
    </div>
  );
};

PlayerScreen.propTypes = {
  id: PropTypes.string.isRequired,
  film: FilmCardType.isRequired,
  loadDataFilm: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  film: getCurrentFilm(state),
});

const mapDispatchToProps = (dispatch) => ({
  loadDataFilm(id) {
    dispatch(fetchCurrentFilm(id));
  },
});

export {PlayerScreen};
export default connect(mapStateToProps, mapDispatchToProps)(PlayerScreen);
