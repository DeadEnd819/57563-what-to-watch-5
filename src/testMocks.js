export const promoFilm = {
  id: 1,
  name: `The Grand Budapest Hotel`,
  posterImage: `the-grand-budapest-hotel-poster.jpg`,
  backgroundImage: `bg-the-grand-budapest-hotel.jpg`,
  backgroundColor: `#ffffff`,
  genre: `Comedy`,
  released: 2014,
  isFavorite: false,
};

export const film = {
  id: 1,
  name: `The Grand Budapest Hotel`,
  posterImage: `the-grand-budapest-hotel-poster.jpg`,
  previewImage: `bg-the-grand-budapest-hotel.jpg`,
  backgroundImage: `bg-the-grand-budapest-hotel.jpg`,
  backgroundColor: `#ffffff`,
  previewVideoLink: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.
  Gustave prides himself on providing first-class service to the hotel's guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.`,
  rating: 8.9,
  scoresCount: 240,
  director: `Wes Andreson`,
  starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Saoirse Ronan`],
  runTime: 99,
  genre: `Comedy`,
  released: 2014,
  isFavorite: true,
};

export const reviewOne = {
  id: 1,
  user: {
    id: 4,
    name: `Kate Muir`,
  },
  rating: 8.9,
  comment: `Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.`,
  date: `2019-05-08T14:13:56.569Z`,
};

export const reviewTwo = {
  id: 2,
  user: {
    id: 5,
    name: `Paula Fleri-Soler`,
  },
  rating: 9.1,
  comment: `The mannered, madcap proceedings are often delightful, occasionally silly, and here and there, gruesome and/or heartbreaking.`,
  date: `2019-08-08T17:40:56.569Z`,
};

export const reviews = [reviewOne, reviewTwo];

export const genresList = [
  `Action`,
  `Adventure`,
  `Comedy`,
  `Crime`,
  `Drama`,
  `Fantasy`,
  `Thriller`
];

export const state = {
  DATA: {
    films: [film, film, film, film, film, film, film],
    promo: promoFilm,
    currentFilm: film,
    currentReviews: reviews,
  },
  FILTER: {
    genre: `Drama`,
    showFilmsCount: 8,
  },
  USER: {
    name: `name`,
    avatar: `avatar.jpg`,
    status: `NOT_AUTHORIZED`,
    reviewStatus: `REVIEW_NOT_UPDATED`,
  },
};
