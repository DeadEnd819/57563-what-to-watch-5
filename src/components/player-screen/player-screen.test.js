import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";
import {PlayerScreen} from "./player-screen";
import {film} from "../../testMocks";

jest.mock(`../player-button/player-button.jsx`, () => `PlayerButton`);
jest.mock(`../full-screen-button/full-screen-button.jsx`, () => `FullScreenButton`);

const renderer = new ShallowRenderer();
const noop = ()=>{};

it(`Should PlayerScreen render correctly`, () => {
  renderer.render(
      <PlayerScreen
        id = {`1`}
        film={film}
        loadDataFilm = {noop}
      />
  );

  const tree = renderer.getRenderOutput();

  expect(tree).toMatchSnapshot();
});
