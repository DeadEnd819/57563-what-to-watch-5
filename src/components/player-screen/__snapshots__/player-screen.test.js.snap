// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should PlayerScreen render correctly 1`] = `
<div
  className="player"
>
  <video
    className="player__video"
    poster="bg-the-grand-budapest-hotel.jpg"
  />
  <Link
    className="player__exit"
    to="/films/1"
    type="button"
  >
    Exit
  </Link>
  <div
    className="player__controls"
  >
    <div
      className="player__controls-row"
    >
      <div
        className="player__time"
      >
        <progress
          className="player__progress"
          max={0}
          value={0}
        />
        <div
          className="player__toggler"
          style={
            Object {
              "left": "0%",
            }
          }
        >
          Toggler
        </div>
      </div>
      <div
        className="player__time-value"
      >
        0:00
      </div>
    </div>
    <div
      className="player__controls-row"
    >
      <PlayerButton
        isPlaying={true}
        onTogglePlayButtonClick={[Function]}
      />
      <div
        className="player__name"
      >
        The Grand Budapest Hotel
      </div>
      <FullScreenButton
        onFullScreenButtonClick={[Function]}
      />
    </div>
  </div>
</div>
`;
